components/cdc/src/service.rs

```diff
@@ -205,30 +221,28 @@ where
 {
     type Error = E;
 
-    /// Will block the task if:
-    /// 1) there are ChangeDataEvents not sent out, or buf` has at least 1024 messages,
-    /// and 2) they cannot be successfully flushed down the `inner_sink` immediately.
+    /// returns Ready if and only if (1) the state machine is in Ready state, and (2)
+    /// there are not so many CdcEvents in the buffer.
+    /// If (1) is not satisfied, it means that this is NOT the first poll to this function in preparation to
+    /// start_send, so it must be the result of the task being awaken by a block in poll_flush, so we call
+    /// poll_flush.
     fn poll_ready(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
         let this = self.get_mut();
 
-        if !this.send_buf.is_empty() {
-            ready!(this.poll_flush_unpin(cx))?;
-        }
-
-        if this.buf.is_none() {
-            this.buf = Some(vec![]);
-            return Poll::Ready(Ok(()));
-        }
-
-        // We set the threshold to flush the inner_sink to 32
-        // TODO find a way to determine the best number
-        if this.buf.as_ref().unwrap().len() >= 32 {
-            this.prepare_flush();
-            ready!(this.inner_sink.poll_flush_unpin(cx))?;
-
-            assert!(this.buf.is_none());
-            this.buf = Some(vec![]);
-            return Poll::Ready(Ok(()));
+        match this.state {
+            EventBatcherSinkState::Ready(ref buf) => {
+                if buf.len() >= 2 {
+                    // converts CdcEvent to ChangeDataEvent
+                    this.prepare_flush();
+                    ready!(this.poll_flush_unpin(cx))?;
+                }
+            }
+            EventBatcherSinkState::Sending(_) | EventBatcherSinkState::Flushing => {
```

Comment:
Can we combine the state `Sending` and `Flushing` into one state

Reply:
Since the `Sending` state is associated with a `deque` but the `Flushing` state is not, combining these two will likely reduce code readability.